lib/internal/Magento/Framework/Module/Dir/Reader.php

```diff
@@ -102,6 +102,13 @@ private function getFiles($filename, $subDir = '')
             $path = $directoryRead->getRelativePath($file);
             if ($directoryRead->isExist($path)) {
                 $result[] = $file;
+            } elseif ($checkParent) {
```

Comment:
It is based on assumption of a module structure. It may be true for most of cases, but not for all.
One of the cases may be that a module is located in an arbitrary location outside of Magento root dir. For example, `/home/user/MyMagentoModule`. And the module may not have `composer.json` file. As a result the logic will allow to read a file under `/home/user`.
The assumption may lead to unsecure behavior of the application. It is better to know 100% where `composer.json` file should be and only check whether it exists or not.

Please, check also this discussion https://github.com/magento/magento2/issues/3400